bundles/org.eclipse.ui.ide/src/org/eclipse/ui/views/markers/internal/ContentGeneratorDescriptor.java

```diff
@@ -203,9 +203,25 @@ public Collection<MarkerType> getMarkerTypes() {
 			IConfigurationElement[] markerTypeElements = configurationElement.getChildren(MarkerSupportRegistry.MARKER_TYPE_REFERENCE);
 			for (IConfigurationElement configElement : markerTypeElements) {
 				String elementName = configElement.getAttribute(MarkerSupportInternalUtilities.ATTRIBUTE_ID);
-				MarkerType[] types = MarkerTypesModel.getInstance().getType(elementName).getAllSubTypes();
-				markerTypes.addAll(Arrays.asList(types));
-				markerTypes.add(MarkerTypesModel.getInstance().getType(elementName));
+
+				String application = configElement.getAttribute(MarkerSupportInternalUtilities.APPLICATION) == null
+						? MarkerSupportInternalUtilities.TYPE_AND_SUBTYPE
+						: configElement.getAttribute(MarkerSupportInternalUtilities.APPLICATION);
+
+				switch (application) {
+				case MarkerSupportInternalUtilities.TYPE_ONLY:
+					markerTypes.add(MarkerTypesModel.getInstance().getType(elementName));
+					break;
+				case MarkerSupportInternalUtilities.SUB_TYPES_ONLY:
+					markerTypes.addAll(
+							Arrays.asList(MarkerTypesModel.getInstance().getType(elementName).getAllSubTypes()));
+					break;
+				case MarkerSupportInternalUtilities.TYPE_AND_SUBTYPE:
+				default:
```

Comment:
i wonder that the default does change the previous implementation. please explain.

Reply:
If you look at the change in ContentGeneratorDescriptor.getMarkerTypes you see that the previous implementation was populating types and subtypes. With the change the application attribute is used to determine what to include. If there is no application attribute then the type and subtype are used just like the old implementation.

Reply:
you changed the order of addAll() and add(). probably not a problem - but why? prefer to keep the old code if there is no need to change

Reply:
No reason for the order change. Its swapped back now.